From a9dca76ad89d56682436983408df1f27d322c12b Mon Sep 17 00:00:00 2001
From: Nikita Popov <npopov@redhat.com>
Date: Wed, 20 Dec 2023 14:41:52 +0100
Subject: [PATCH 27/28] [rust] [MergeFunc] Add tests for incorrect const expr
 merging (NFC)

(cherry picked from commit 1ff9fb78c8bb87ffa8700523cd77e4e2bf000740)
(cherry picked from commit 3dd2db08a2b30618e21f165cf094de421dc32c00)
(cherry picked from commit 836e71a4254cb7b61f9d1b59591e69bb4055eb45)
---
 llvm/test/Transforms/MergeFunc/constexpr.ll | 55 +++++++++++++++++++++
 1 file changed, 55 insertions(+)
 create mode 100644 llvm/test/Transforms/MergeFunc/constexpr.ll

diff --git a/llvm/test/Transforms/MergeFunc/constexpr.ll b/llvm/test/Transforms/MergeFunc/constexpr.ll
new file mode 100644
index 000000000000..7a4823b5c494
--- /dev/null
+++ b/llvm/test/Transforms/MergeFunc/constexpr.ll
@@ -0,0 +1,55 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py UTC_ARGS: --check-globals all --version 4
+; RUN: opt -S -passes=mergefunc -mergefunc-use-aliases < %s | FileCheck %s
+
+@g1 = external unnamed_addr global i8
+@g2 = external unnamed_addr global i8
+
+;.
+; CHECK: @g1 = external unnamed_addr global i8
+; CHECK: @g2 = external unnamed_addr global i8
+; CHECK: @f2 = unnamed_addr alias i1 (), ptr @f1
+; CHECK: @f4 = unnamed_addr alias ptr (), ptr @f3
+; CHECK: @f5 = unnamed_addr alias ptr (), ptr @f3
+; CHECK: @f7 = unnamed_addr alias i64 (), ptr @f6
+; CHECK: @f8 = unnamed_addr alias i64 (), ptr @f6
+;.
+define i1 @f1() unnamed_addr {
+; CHECK-LABEL: define i1 @f1() unnamed_addr {
+; CHECK-NEXT:    ret i1 icmp eq (ptr @g1, ptr @g2)
+;
+  ret i1 icmp eq (ptr @g1, ptr @g2)
+}
+
+define i1 @f2() unnamed_addr {
+  ret i1 icmp ne (ptr @g1, ptr @g2)
+}
+
+define ptr @f3() unnamed_addr {
+; CHECK-LABEL: define ptr @f3() unnamed_addr {
+; CHECK-NEXT:    ret ptr getelementptr inbounds (i8, ptr @g1, i64 2)
+;
+  ret ptr getelementptr inbounds (i8, ptr @g1, i64 2)
+}
+
+define ptr @f4() unnamed_addr {
+  ret ptr getelementptr (i16, ptr @g1, i64 2)
+}
+
+define ptr @f5() unnamed_addr {
+  ret ptr getelementptr (i8, ptr @g1, i64 2)
+}
+
+define i64 @f6() unnamed_addr {
+; CHECK-LABEL: define i64 @f6() unnamed_addr {
+; CHECK-NEXT:    ret i64 add nuw (i64 ptrtoint (ptr @g1 to i64), i64 1)
+;
+  ret i64 add nuw (i64 ptrtoint (ptr @g1 to i64), i64 1)
+}
+
+define i64 @f7() unnamed_addr {
+  ret i64 add (i64 ptrtoint (ptr @g1 to i64), i64 1)
+}
+
+define i64 @f8() unnamed_addr {
+  ret i64 sub (i64 ptrtoint (ptr @g1 to i64), i64 1)
+}
-- 
2.43.0

